### PR TITLE
[SPARK-31037][SQL] refine AQE config names

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

When introducing AQE to others, I feel the config names are a bit incoherent and hard to use.
This PR refines the config names:
1. remove the "shuffle" prefix. AQE is all about shuffle and we don't need to add the "shuffle" prefix everywhere.
2. `targetPostShuffleInputSize` is obscure, rename to `advisoryPartitionSizeInBytes`.
3. `reducePostShufflePartitions` doesn't match the actual optimization, rename to `coalescePartitions`
4. `minNumPostShufflePartitions` is obscure, rename it `minPartitionNum` under the `coalescePartitions` namespace
5. `maxNumPostShufflePartitions` is confusing with the word "max", rename it `initialPartitionNum`
6. `skewedJoinOptimization` is too verbose. skew join is a well-known terminology in database area, we can just say `skewJoin` 

### Why are the changes needed?

Make the config names easy to understand.

### Does this PR introduce any user-facing change?

deprecate the config `spark.sql.adaptive.shuffle.targetPostShuffleInputSize`

### How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @JkSelf @maryannxue

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -388,99 +398,97 @@ object SQLConf {
...
-  val REDUCE_POST_SHUFFLE_PARTITIONS_ENABLED =
-    buildConf("spark.sql.adaptive.shuffle.reducePostShufflePartitions")
-      .doc(s"When true and '${ADAPTIVE_EXECUTION_ENABLED.key}' is enabled, this enables reducing " +
-        "the number of post-shuffle partitions based on map output statistics.")
+  val ADVISORY_SHUFFLE_PARTITION_SIZE_IN_BYTES =
+    buildConf("spark.sql.adaptive.advisoryShufflePartitionSizeInBytes")
+      .doc("The advisory size in bytes of the shuffle partition during adaptive optimization. " +
+        "It takes effect when Spark coalesces small shuffle partitions or splits skewed shuffle " +
+        "partition.")
+      .version("3.0")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `3.0.0`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -388,99 +398,97 @@ object SQLConf {
...
-    buildConf("spark.sql.adaptive.shuffle.fetchShuffleBlocksInBatch")
-      .doc("Whether to fetch the continuous shuffle blocks in batch. Instead of fetching blocks " +
-        "one by one, fetching continuous shuffle blocks for the same map task in batch can " +
-        "reduce IO and improve performance. Note, multiple continuous blocks exist in single " +
-        s"fetch request only happen when '${ADAPTIVE_EXECUTION_ENABLED.key}' and " +
-        s"'${REDUCE_POST_SHUFFLE_PARTITIONS_ENABLED.key}' is enabled, this feature also depends " +
-        "on a relocatable serializer, the concatenation support codec in use and the new version " +
-        "shuffle fetch protocol.")
+  val COALESCE_SHUFFLE_PARTITIONS_ENABLED =
+    buildConf("spark.sql.adaptive.coalesceShufflePartitions.enable")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `enabled`

Comment (maryannxue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: enable -> "enabled"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -388,99 +398,97 @@ object SQLConf {
...
+  val COALESCE_SHUFFLE_PARTITIONS_MIN_PARTITION_NUM =
+    buildConf("spark.sql.adaptive.coalesceShufflePartitions.minPartitionNum")
+      .doc("The minimum number of shuffle partitions after coalescing. This configuration only " +
+        s"has an effect when '${ADAPTIVE_EXECUTION_ENABLED.key}' and " +
+        s"'${COALESCE_SHUFFLE_PARTITIONS_ENABLED.key}' are both true.")
       .version("3.0.0")
       .intConf
-      .checkValue(_ > 0, "The minimum shuffle partition number " +
-        "must be a positive integer.")
+      .checkValue(_ > 0, "The number of partitions must be positive.")
```
Comment (maryannxue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: "number" -> "minimum number"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -388,99 +397,98 @@ object SQLConf {
...
-    buildConf("spark.sql.adaptive.shuffle.fetchShuffleBlocksInBatch")
-      .doc("Whether to fetch the continuous shuffle blocks in batch. Instead of fetching blocks " +
-        "one by one, fetching continuous shuffle blocks for the same map task in batch can " +
-        "reduce IO and improve performance. Note, multiple continuous blocks exist in single " +
-        s"fetch request only happen when '${ADAPTIVE_EXECUTION_ENABLED.key}' and " +
-        s"'${REDUCE_POST_SHUFFLE_PARTITIONS_ENABLED.key}' is enabled, this feature also depends " +
-        "on a relocatable serializer, the concatenation support codec in use and the new version " +
-        "shuffle fetch protocol.")
+  val COALESCE_SHUFFLE_PARTITIONS_ENABLED =
+    buildConf("spark.sql.adaptive.coalesceShufflePartitions.enabled")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Can we make these shorter and easier to remember? Personally,  maybe change `spark.sql.adaptive.coalesceShufflePartitions.enabled` to `spark.sql.adaptive.mergePartitions`, `spark.sql.adaptive.coalesceShufflePartitions.initialPartitionNum` to `spark.sql.adaptive.mergePartitions.initialNum` and `spark.sql.adaptive.coalesceShufflePartitions.minPartitionNum` to `spark.sql.adaptive.mergePartiontions.minNum`.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Shuffle is where this behavior happens and have been written in the doc field, we may not need to enforce it in the config name, and it seems that we do not have any other places under `adaptive` to coalesce partitions. And merge might be easier to spell than coalesce :)

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I'm fine with "coalescePartitions".

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: shall we also rename `advisoryShufflePartitionSizeInBytes` to `advisoryPartitionSizeInBytes`?

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `advisoryPartitionSizeInBytes` looks good to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -388,99 +397,98 @@ object SQLConf {
...
-    buildConf("spark.sql.adaptive.shuffle.localShuffleReader.enabled")
-    .doc(s"When true and '${ADAPTIVE_EXECUTION_ENABLED.key}' is enabled, this enables the " +
-      "optimization of converting the shuffle reader to local shuffle reader for the shuffle " +
-      "exchange of the broadcast hash join in probe side.")
-    .version("3.0.0")
-    .booleanConf
-    .createWithDefault(true)
+    buildConf("spark.sql.adaptive.localShuffleReader.enabled")
+      .doc(s"When true and '${ADAPTIVE_EXECUTION_ENABLED.key}' is true, Spark tries to use local " +
+        "shuffle reader to read the shuffle data of the probe side of a broadcast-hash join.")
```
Comment (JkSelf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: local shuffle reader may optimize the local reader both build side and probe side?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (JkSelf) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Can we uniform the verb naming in `reduce`, `coalesce`, or `merge `both in here configuration name and the optimization rule name  of `ReduceNumShufflePartitions` ? If we use the `coalescePartitions`, It is better to modify the optimization rule name from `ReduceNumShufflePartitions`  to `CoalesceShufflePartitions` ?

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Think we should, but we can do it in another PR. The code name is not user facing and doesn't need to made into 3.0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 6]
```diff
@@ -67,8 +67,8 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
    * SHUFFLE_TARGET_POSTSHUFFLE_INPUT_SIZE.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: ?change it to ADVISORY_PARTITION_SIZE_IN_BYTES?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -378,8 +378,17 @@ object SQLConf {
     .checkValue(_ > 0, "The value of spark.sql.shuffle.partitions must be positive")
     .createWithDefault(200)
 
+  val SHUFFLE_TARGET_POSTSHUFFLE_INPUT_SIZE =
+    buildConf("spark.sql.adaptive.shuffle.targetPostShuffleInputSize")
+      .internal()
+      .doc("(Deprecated since Spark 3.0)")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Also tell users what is the new conf that replaces it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -388,99 +397,99 @@ object SQLConf {
...
     .booleanConf
     .createWithDefault(false)
 
-  val REDUCE_POST_SHUFFLE_PARTITIONS_ENABLED =
-    buildConf("spark.sql.adaptive.shuffle.reducePostShufflePartitions")
-      .doc(s"When true and '${ADAPTIVE_EXECUTION_ENABLED.key}' is enabled, this enables reducing " +
-        "the number of post-shuffle partitions based on map output statistics.")
+  val ADVISORY_PARTITION_SIZE_IN_BYTES =
+    buildConf("spark.sql.adaptive.advisoryPartitionSizeInBytes")
+      .doc("The advisory size in bytes of the shuffle partition during adaptive optimization. " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: > The advisory size in bytes of the shuffle partition during adaptive optimization (when '${ADAPTIVE_EXECUTION_ENABLED.key}' is true).

**Review by gatorsmile — APPROVED** [review]
LGTM except a few minor comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: The doc here is not user-facing. End users will see a message to suggest the new config by https://github.com/apache/spark/pull/27793/files#diff-9a6b543db706f1a90f790783d6930a13R2494

**Comment by cloud-fan** [comment]
thanks for review, merging to master/3.0!

**Pull request closed without merging**